debugger/src/main.rs

```diff
@@ -0,0 +1,184 @@
+//     src.parse()
+// }
+
+///CliCommands enum contains the list of supported commands that can be invoked from the command line.
+///The `debug` command is also the entry point for the DAP server, when the --dap option is used.
+#[derive(StructOpt)]
+#[structopt(
+    name = crate_name!(),
+    about = crate_description!(),
+    author = crate_authors!(),
+    version = crate_version!()
+)]
+enum CliCommands {
+    /// List all connected debug probes
+    #[structopt(name = "list")]
+    List {},
+    /// Gets infos about the selected debug probe and connected target
+    #[structopt(name = "info")]
+    Info {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+    },
+    /// Resets the target attached to the selected debug probe
+    #[structopt(name = "reset")]
+    Reset {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+
+        /// Whether the reset pin should be asserted or deasserted. If left open, just pulse it
+        assert: Option<bool>,
+    },
+    ///Open target in debug mode and accept debug commands.
+    ///By default, the program operates in CLI mode.
+    #[structopt(name = "debug")]
+    Debug {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+
+        //TODO: Implement multi-session --server choices
+        /// Switch from using CLI to DAP Protocol debug commands. By default, the DAP communication for the first session is via STDIN and STDOUT. Adding the additional --port property will run as an IP server, listening to connections on the specified port.
+        #[structopt(long)]
+        dap: bool,
+    },
+    /// Dump memory from attached target
+    #[structopt(name = "dump")]
+    Dump {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+
+        /// The address of the memory to dump from the target (in hexadecimal without 0x prefix)
+        #[structopt(parse(try_from_str = parse_hex))]
+        loc: u32,
+        /// The amount of memory (in words) to dump
+        words: u32,
+    },
+    /// Download memory to attached target
+    #[structopt(name = "download")]
+    Download {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+
+        /// The path to the file to be downloaded to the flash
+        path: String,
+    },
+    ///Begin tracing a memory address over SWV
+    #[structopt(name = "trace")]
+    Trace {
+        #[structopt(flatten)]
+        debugger_options: DebuggerOptions,
+
+        /// The address of the memory start trace (in hexadecimal without 0x prefix)
+        #[structopt(parse(try_from_str = parse_hex))]
+        loc: u32,
+    },
+}
+fn main() -> Result<()> {
+    // Initialize the logging backend.
+    pretty_env_logger::formatted_builder()
+        .target(pretty_env_logger::env_logger::Target::Stderr) // Log to Stderr, because the DebugAdapater, in 'Launch' mode, needs Stdin and Stdout to communicate with VSCode DAP Client
+        .init();
```

Comment:
This seems to ignore the `RUST_LOG` environment variable. How can I configure which level is logged?

Reply:
Thanks for catching this. I am replacing pretty_env_logger with the underlying env_logger. It is leaner and I don't need all the features. In an ideal world, I'd like to find a way to direct all the RUST logs to the `debug_adapter::log_to_console()` method, but I need help with that ... because the debug_adapter is only initialized late in the startup. Any ideas?

Reply:
You can have a look at https://github.com/probe-rs/probe-rs/blob/master/probe-rs-cli-util/src/logging.rs for ideas, @Yatekii should know more about it. Another solution can also be to just log into a file.

I'm also fine with improving this later, just being able to select the log level would be good enough for me. With the TCP server console logs work quite well.

Reply:
Excellent. I will add a TODO in the code. Once the PR has been promoted, I will systematically add Issues/PR's for individual TODO's.